host/bcm283x/dma.go

```diff
@@ -734,28 +751,36 @@ func dmaWriteStreamPCM(p *Pin, w gpiostream.Stream) error {
 	}
 
 	_, waits, err := setPCMClockSource(hz)
+	waits = 0
 	if err != nil {
 		return err
 	}
 	pcmMemory.reset()
 	reg := pcmBaseAddr + 0x4 // pcmMap.fifo
-	if err := cb.initBlock(uint32(buf.PhysAddr()), reg, uint32(len(bits.Bits)), false, true, dmaPCMTX, waits); err != nil {
+	if err := cb.initBlock(uint32(buf.PhysAddr()), reg, uint32(len(bits.Bits)), false, true, true, false, dmaPCMTX, waits); err != nil {
 		return err
 	}
 
 	// Try using a full bandwidth channel.
-	chNum, ch := pickChannel(6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
+	_, ch := pickChannel(6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
 	if ch == nil {
 		return errors.New("bcm283x-dma: no channel available")
 	}
+	pcmMemory.set()
+
 	defer ch.reset()
-	log.Printf("Channel: %d", chNum)
 	ch.startIO(uint32(pCB.PhysAddr()))
 
-	pcmMemory.set()
 	defer pcmMemory.reset()
-	log.Printf("PCM: %#v", pcmMemory)
-	return ch.wait()
+
+	if err := ch.wait(); err != nil {
+		return err
+	}
+	// We have to wait PCM to be finished even after DMA finished.
+	for pcmMemory.cs&pcmTXEmpty == 0 {
```

Comment:
This does not seem enough. Even after FIFO is empty, PCM could still sending remaining bits. I sometimes observed the last led is not updated correctly because we reset pcmMemory too early.

Reply:
Looks like we should wait for pcmTXErr.

Reply:
Any idea why?

Reply:
The problem is that we call reset() and it disables PCM even in the middle of transmission.
Even after DMA write finishes, data is still in FIFO. Even after FIFO become empty, bits in the last word could be still in the transmission queue. I guess pcmTXErr raised when there is not bits to be transmitted.